File: 2026-08-15-csv-download-with-streaming/action-in-blog/src/main/java/action/in/blog/todo/service/TodoService.java
package action.in.blog.todo.service;

import action.in.blog.todo.repository.TodoRepository;
import jakarta.persistence.EntityManager;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

@Service
public class TodoService {

    private final TodoRepository todoRepository;
    private final EntityManager entityManager;

    public TodoService(TodoRepository todoRepository, EntityManager entityManager) {
        this.todoRepository = todoRepository;
        this.entityManager = entityManager;
    }

    @Transactional(readOnly = true)
    public void exportCsv(OutputStream outputStream) {
        try (var stream = todoRepository.findAllAsStream()) {
            AtomicInteger count = new AtomicInteger();
            var csvHeader = "id,title,description,completed\n";
            outputStream.write(csvHeader.getBytes(StandardCharsets.UTF_8));
            stream.forEach(todo -> {
                var row = String.join(",",
                        List.of(
                                todo.getId().toString(),
                                todo.getTitle(),
                                todo.getDescription(),
                                todo.isCompleted() ? "완료" : "미완료",
                                "\n"
                        )
                );
                count.getAndIncrement();
                try {
                    outputStream.write(row.getBytes(StandardCharsets.UTF_8));
                    if (count.get() % 1000 == 0) {
                        outputStream.flush();
                        entityManager.clear();
                    }
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
            outputStream.write("\uFEFF".getBytes(StandardCharsets.UTF_8));
            outputStream.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: 2024-08-07-aws-ecr-setup/action-in-blog/backend/src/main/kotlin/com/example/demo/service/TodoService.kt
package com.example.demo.service

import com.example.demo.domain.Todo
import com.example.demo.domain.TodoEntity
import com.example.demo.repository.TodoRepository
import com.example.demo.util.TimeProvider.parseStringToTimestamp
import com.example.demo.util.TimeProvider.toStringFormat
import org.springframework.stereotype.Service

interface TodoService {
    fun getTodos(): List<Todo>
    fun createTodo(todo: Todo)
}

@Service
class DefaultTodoService(
    private val todoRepository: TodoRepository
) : TodoService {
    override fun getTodos(): List<Todo> {
        return todoRepository.findAllByOrderByCreatedAtDesc()
            .map {
                Todo(
                    id = it.id,
                    title = it.title,
                    createdAt = toStringFormat(it.createdAt)
                )
            }
    }

    override fun createTodo(todo: Todo) {
        todoRepository.save(
            TodoEntity(
                title = todo.title,
                createdAt = parseStringToTimestamp(todo.createdAt)
            )
        )
    }
}